File: src/graph/dfs/km103.py
from typing import List, Tuple


step: List[Tuple[int, int]] = [(1, 0), (0, 1), (-1, 0), (0, -1)]

grid: List[List[int]] = []
luVisit: List[List[bool]]
rdVisit: List[List[bool]]

rowNum, colNum = map(int, input().strip().split(" "))

for _ in range(rowNum):
    grid.append(list(map(int, input().strip().split(" "))))

luVisit = [[False for _ in range(colNum)] for _ in range(rowNum)]
rdVisit = [[False for _ in range(colNum)] for _ in range(rowNum)]

ans: List[Tuple[int, int]] = []


def rico() -> None:

    def dfs(y: int, x: int, visit: List[List[bool]]) -> None:
        visit[y][x] = True

        for item in step:
            dy, dx = item
            nextY, nextX = y + dy, x + dx
            if (
                nextY >= 0
                and nextY < rowNum
                and nextX >= 0
                and nextX < colNum
                and grid[y][x] <= grid[nextY][nextX]
                and (not visit[nextY][nextX])
            ):
                dfs(nextY, nextX, visit)

    for x in range(colNum):
        dfs(0, x, luVisit)
        dfs(rowNum - 1, x, rdVisit)

    for y in range(rowNum):
        dfs(y, 0, luVisit)
        dfs(y, colNum - 1, rdVisit)

    for y in range(rowNum):
        for x in range(colNum):
            if luVisit[y][x] and rdVisit[y][x]:
                ans.append((y, x))


rico()
for item in ans:
    print(item[0], item[1])

File: src/graph/dfs/km103.cc
#include <functional>
#include <iostream>
#include <vector>

using namespace std;

auto step = vector<pair<int, int>>{{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

int rowNum;
int colNum;

vector<vector<int>> grid;
vector<vector<bool>> luVisit;
vector<vector<bool>> rdVisit;

int main() {
  cin >> rowNum >> colNum;
  grid = vector<vector<int>>(rowNum, vector<int>(colNum, 0));

  for (auto y = 0; y < rowNum; y++) {
    for (auto x = 0; x < colNum; x++) {
      cin >> grid[y][x];
    }
  }

  luVisit = vector<vector<bool>>(rowNum, vector<bool>(colNum, false));
  rdVisit = vector<vector<bool>>(rowNum, vector<bool>(colNum, false));

  function<void(int, int, vector<vector<bool>>&)> dfs;

  dfs = [&](int y, int x, vector<vector<bool>>& visit) -> void {
    visit[y][x] = true;

    for (const auto& item : step) {
      auto dy = item.first;
      auto dx = item.second;
      auto nextY = y + dy;
      auto nextX = x + dx;
      if (nextY >= 0 && nextY < rowNum && nextX >= 0 && nextX < colNum &&
          grid[y][x] <= grid[nextY][nextX] && !visit[nextY][nextX]) {
        dfs(nextY, nextX, visit);
      }
    }
  };

  for (auto x = 0; x < colNum; x++) {
    dfs(0, x, luVisit);
    dfs(rowNum - 1, x, rdVisit);
  }

  for (auto y = 0; y < rowNum; y++) {
    dfs(y, 0, luVisit);
    dfs(y, colNum - 1, rdVisit);
  }

  for (auto y = 0; y < rowNum; y++) {
    for (auto x = 0; x < colNum; x++) {
      if (luVisit[y][x] && rdVisit[y][x]) {
        cout << y << " " << x << endl;
      }
    }
  }
}
